File: bindings/node/tests/test_download.test.js
import assert from "node:assert/strict";
import { test } from "node:test";
import { TikTokExtractor, download } from "../src/index.js";
import { join } from "node:path";
import { rm } from "node:fs/promises";
import { existsSync, statSync, mkdirSync } from "node:fs";

const URL = "https://vt.tiktok.com/ZSxvYRvoR/";

test("test_invalid_download_params_raise_error", async () => {
  const extractor = new TikTokExtractor();
  await assert.rejects(
    async () => await extractor.download(URL, "invalid-kind", "downloads")
  );

  await assert.rejects(
    async () => await download(URL, "invalid-kind", "downloads")
  );
});

test("test_download_real_url", async () => {
  const tempDir = join(process.cwd(), "../../temp");
  const outDir = join(tempDir, `tiktok-direct-node-${Date.now()}`);
  mkdirSync(outDir, { recursive: true });
  
  const extractor = new TikTokExtractor();
  
  try {
    const savedPath = await extractor.download(URL, "mp4", outDir);
    assert.ok(existsSync(savedPath));
    if (statSync(savedPath).size === 0) {
      throw new Error("media download error: CDN returned 0 bytes");
    }
  } catch (err) {
    if (!err.message.includes("no media URL is available") && !err.message.includes("media download error") && !err.message.includes("all media candidates failed")) {
      throw err;
    }
  } finally {
    await rm(outDir, { recursive: true, force: true });
  }
});

File: bindings/go/tests/test_download_test.go
package tests

import (
	"context"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"testing"
	"time"

	tiktokdirect "github.com/codewithwan/tiktok-direct-go"
)

func TestInvalidDownloadParamsRaiseError(t *testing.T) {
	extractor := tiktokdirect.New()
	_, err := extractor.Download(context.Background(), URL, "invalid-kind", "downloads")
	if err == nil {
		t.Error("Expected error for invalid kind, got nil")
	}

	_, err = tiktokdirect.Download(context.Background(), URL, "invalid-kind", "downloads")
	if err == nil {
		t.Error("Expected error for invalid kind in package function, got nil")
	}
}

func TestDownloadRealUrl(t *testing.T) {
	outDir := filepath.Join("..", "..", "..", "temp", fmt.Sprintf("tiktok-direct-go-%d", time.Now().UnixNano()))
	err := os.MkdirAll(outDir, 0755)
	if err != nil {
		t.Fatalf("Failed to create temp dir: %v", err)
	}
	defer os.RemoveAll(outDir)

	extractor := tiktokdirect.New()
	path, err := extractor.Download(context.Background(), URL, "mp4", outDir)
	
	if err != nil {
		errStr := err.Error()
		if !strings.Contains(errStr, "no media URL") && !strings.Contains(errStr, "all media candidates failed") {
			t.Fatalf("Unexpected download error: %v", err)
		}
	} else {
		info, err := os.Stat(path)
		if err != nil {
			t.Fatalf("File not created: %v", err)
		}
		if info.Size() == 0 {
			t.Error("Expected non-empty file")
		}
	}
}
